File: TS/系统/03．技能系统/05．单位技能/03．Boss技能/01．主线Boss/06．菲利斯/01．运行时上下文.ts
/** @noSelfInFile */

import type { 机制清理篮子 } from "../../../../00．技能模板+函数/04．机制组件/06．机制清理";
import { 创建单位运行时上下文工厂 } from "../../../../00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/15．单位运行时上下文工厂";
import { 菲利斯单位技能配置 } from "./00．配置";
import { 播放菲利斯台词 } from "./08．台词播放";

const { getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  getServerTime: (this: void) => number;
};

export type 菲利斯阶段 = 1 | 2;

export interface 菲利斯剑魂狼记录 {
  Boss单位: any;
  大狼: boolean;
  伤害比例: number;
}

export interface 菲利斯运行时上下文 {
  Boss单位: any;
  阶段: 菲利斯阶段;
  开战时间Ms: number;
  清理: 机制清理篮子;
  当前魔法充能: number;
  当前领袖光环低血: boolean;
  领袖光环清理已注册: boolean;
  异形化准备中: boolean;
  异形化中: boolean;
  异形化结束Ms: number;
  已初始化: boolean;
}

const 菲利斯剑魂狼表: Record<number, 菲利斯剑魂狼记录 | undefined> = {};

function 创建菲利斯上下文(this: void, boss: any, 清理: 机制清理篮子): 菲利斯运行时上下文 {
  播放菲利斯台词(boss, "开场", 0);
  return {
    Boss单位: boss,
    阶段: 1,
    开战时间Ms: getServerTime(),
    清理,
    当前魔法充能: 0,
    当前领袖光环低血: false,
    领袖光环清理已注册: false,
    异形化准备中: false,
    异形化中: false,
    异形化结束Ms: 0,
    已初始化: false,
  };
}

function on菲利斯单位死亡(this: void, _context: 菲利斯运行时上下文, dyingUnit: any, _killingUnit: any): void {
  播放菲利斯台词(dyingUnit, "死亡", 0);
}

const 菲利斯上下文工厂 = 创建单位运行时上下文工厂<菲利斯运行时上下文>({
  名称: "菲利斯",
  主动技能提示: 菲利斯单位技能配置.主动技能提示,
  创建上下文: 创建菲利斯上下文,
  死亡时自动清理: true,
  on单位死亡: on菲利斯单位死亡,
});

export function 获取菲利斯上下文(this: void, boss: any): 菲利斯运行时上下文 | undefined {
  return 菲利斯上下文工厂.获取(boss);
}

export function 获取或创建菲利斯上下文(this: void, boss: any): 菲利斯运行时上下文 | undefined {
  return 菲利斯上下文工厂.获取或创建(boss);
}

export function 清理菲利斯上下文(this: void, boss: any): void {
  菲利斯上下文工厂.清理上下文(boss);
}

export function 获取全部菲利斯上下文(this: void): 菲利斯运行时上下文[] {
  return 菲利斯上下文工厂.获取全部();
}

export function 登记菲利斯剑魂狼(this: void, wolf: any, record: 菲利斯剑魂狼记录): void {
  const id = 菲利斯上下文工厂.取单位ID(wolf);
  if (id === 0) return;
  菲利斯剑魂狼表[id] = record;
}

export function 注销菲利斯剑魂狼(this: void, wolf: any): void {
  const id = 菲利斯上下文工厂.取单位ID(wolf);
  if (id !== 0) delete 菲利斯剑魂狼表[id];
}

export function 获取菲利斯剑魂狼记录(this: void, wolf: any): 菲利斯剑魂狼记录 | undefined {
  const id = 菲利斯上下文工厂.取单位ID(wolf);
  return id === 0 ? undefined : 菲利斯剑魂狼表[id];
}

File: src/系统/03．技能系统/05．单位技能/03．Boss技能/01．主线Boss/06．菲利斯/01．运行时上下文.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
local ____15_FF0E_5355_4F4D_8FD0_884C_65F6_4E0A_4E0B_6587_5DE5_5382 = require("系统.03．技能系统.00．技能模板+函数.04．机制组件.10．复杂战斗通用机制.15．单位运行时上下文工厂")
local _____521B_5EFA_5355_4F4D_8FD0_884C_65F6_4E0A_4E0B_6587_5DE5_5382 = ____15_FF0E_5355_4F4D_8FD0_884C_65F6_4E0A_4E0B_6587_5DE5_5382["创建单位运行时上下文工厂"]
local ____00_FF0E_914D_7F6E = require("系统.03．技能系统.05．单位技能.03．Boss技能.01．主线Boss.06．菲利斯.00．配置")
local _____83F2_5229_65AF_5355_4F4D_6280_80FD_914D_7F6E = ____00_FF0E_914D_7F6E["菲利斯单位技能配置"]
local ____08_FF0E_53F0_8BCD_64AD_653E = require("系统.03．技能系统.05．单位技能.03．Boss技能.01．主线Boss.06．菲利斯.08．台词播放")
local _____64AD_653E_83F2_5229_65AF_53F0_8BCD = ____08_FF0E_53F0_8BCD_64AD_653E["播放菲利斯台词"]
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local getServerTime = ____require_result_0.getServerTime
local _____83F2_5229_65AF_5251_9B42_72FC_8868 = {}
local function _____521B_5EFA_83F2_5229_65AF_4E0A_4E0B_6587(boss, _____6E05_7406)
    _____64AD_653E_83F2_5229_65AF_53F0_8BCD(boss, "开场", 0)
    return {
        ["Boss单位"] = boss,
        ["阶段"] = 1,
        ["开战时间Ms"] = getServerTime(),
        ["清理"] = _____6E05_7406,
        ["当前魔法充能"] = 0,
        ["当前领袖光环低血"] = false,
        ["原生领袖光环已移除"] = false,
        ["领袖光环清理已注册"] = false,
        ["异形化中"] = false,
        ["异形化结束Ms"] = 0,
        ["已初始化"] = false
    }
end
local function ____on_83F2_5229_65AF_5355_4F4D_6B7B_4EA1(_context, dyingUnit, _killingUnit)
    _____64AD_653E_83F2_5229_65AF_53F0_8BCD(dyingUnit, "死亡", 0)
end
local _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382 = _____521B_5EFA_5355_4F4D_8FD0_884C_65F6_4E0A_4E0B_6587_5DE5_5382({
    ["名称"] = "菲利斯",
    ["主动技能提示"] = _____83F2_5229_65AF_5355_4F4D_6280_80FD_914D_7F6E["主动技能提示"],
    ["创建上下文"] = _____521B_5EFA_83F2_5229_65AF_4E0A_4E0B_6587,
    ["死亡时自动清理"] = true,
    ["on单位死亡"] = ____on_83F2_5229_65AF_5355_4F4D_6B7B_4EA1
})
____exports["获取菲利斯上下文"] = function(boss)
    return _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["获取"](boss)
end
____exports["获取或创建菲利斯上下文"] = function(boss)
    return _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["获取或创建"](boss)
end
____exports["清理菲利斯上下文"] = function(boss)
    _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["清理上下文"](boss)
end
____exports["获取全部菲利斯上下文"] = function()
    return _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["获取全部"]()
end
____exports["登记菲利斯剑魂狼"] = function(wolf, record)
    local id = _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["取单位ID"](wolf)
    if id == 0 then
        return
    end
    _____83F2_5229_65AF_5251_9B42_72FC_8868[id] = record
end
____exports["注销菲利斯剑魂狼"] = function(wolf)
    local id = _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["取单位ID"](wolf)
    if id ~= 0 then
        __TS__Delete(_____83F2_5229_65AF_5251_9B42_72FC_8868, id)
    end
end
____exports["获取菲利斯剑魂狼记录"] = function(wolf)
    local id = _____83F2_5229_65AF_4E0A_4E0B_6587_5DE5_5382["取单位ID"](wolf)
    local ____temp_1
    if id == 0 then
        ____temp_1 = nil
    else
        ____temp_1 = _____83F2_5229_65AF_5251_9B42_72FC_8868[id]
    end
    return ____temp_1
end
return ____exports
